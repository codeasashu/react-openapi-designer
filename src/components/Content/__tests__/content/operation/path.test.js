import React from 'react';
import {
  ContentUtils,
  render,
  userEvent,
  initStore,
} from '../../../../../../test-utils';
import Path from '../../../path';

let _operationPath, _operation;

const operationTestId = 'path-operation-111';
const otherPropertiesTestId = 'other-properties-222';

jest.mock(
  '../../../operation/operation',
  // eslint-disable-next-line react/display-name
  () => (e) => {
    _operation = e.operation;
    _operationPath = e.relativeJsonPath;
    return <div data-testid={operationTestId}>def</div>;
  },
);

let _parameterPath;

jest.mock(
  '../../../../Editor/JsonSchema/OtherProperties/parameters',
  // eslint-disable-next-line react/display-name
  () => (e) => {
    _parameterPath = e.relativeJsonPath;
    return <div data-testid={otherPropertiesTestId}>def</div>;
  },
);

describe('Path tests', () => {
  const user = userEvent.setup();
  let store, relativeJsonPath, rerender_node, _getParsedOasData;

  const re_renderer = (store, node, rerender) => {
    return () => {
      rerender(<Path relativeJsonPath={node.relativeJsonPath} node={node} />, {
        providerProps: {value: store},
      });
      return node;
    };
  };

  beforeEach(async () => {
    const {stores, findNode, getData} = initStore({
      paths: {'/user/abc': {post: {summary: 'Hello'}}},
    });
    const node = findNode.path('~1user~1abc/post');
    stores.uiStore.setActiveNode(node);
    relativeJsonPath = node.relativeJsonPath;
    const {rerender} = render(
      <Path relativeJsonPath={relativeJsonPath} node={node} />,
      {
        providerProps: {value: stores},
      },
    );
    rerender_node = re_renderer(store, node, rerender);
    store = stores;
    _getParsedOasData = getData;
  });

  it('has valid methods', async () => {
    expect(ContentUtils.path().methodTabs()).toHaveLength(8);
    expect(ContentUtils.path().methodTab('get')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('post')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('put')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('patch')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('delete')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('head')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('options')).toBeInTheDocument();
    expect(ContentUtils.path().methodTab('trace')).toBeInTheDocument();

    expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');
    expect(ContentUtils.path().pathParamsBtn()).toBeInTheDocument();
    expect(ContentUtils.path().hostnameSelect()).toBeInTheDocument();
    expect(ContentUtils.path().operationNameInput()).toHaveValue('Hello');
    expect(ContentUtils.path().operationTab('post')).toHaveClass('active');

    expect(relativeJsonPath).toStrictEqual(['paths', '/user/abc', 'post']);
    expect(store).toHaveSchemaAt(relativeJsonPath, {
      summary: 'Hello',
    });
  });

  it('can change operation name', async () => {
    await user.clear(ContentUtils.path().operationNameInput(), 'Abc');
    await user.type(ContentUtils.path().operationNameInput(), 'Abc');
    expect(ContentUtils.path().operationNameInput()).toHaveValue('Abc');
    expect(store).toHaveSchemaAt([...relativeJsonPath, 'summary'], 'Abc');
  });

  it('can switch methods', async () => {
    expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');
    let parsedData = _getParsedOasData(relativeJsonPath.slice(0, -1));
    expect(Object.keys(parsedData)).toStrictEqual(['post']);
    await user.click(ContentUtils.path().methodTab('get'));
    await user.click(ContentUtils.path().addOperationBtn('get'));
    // Get method has been added
    expect(ContentUtils.path().activeMethodTab()).toHaveTextContent('get');
    expect(ContentUtils.path().operationTab('get')).toHaveClass('active');
    expect(ContentUtils.path().operationTab('post')).not.toHaveClass('active');
    expect(store).toHaveSchemaAt([...relativeJsonPath.slice(0, -1), 'get'], {
      operationId: 'get-user-abc',
      responses: {200: {description: 'OK'}},
      summary: '',
    });
    parsedData = _getParsedOasData(relativeJsonPath.slice(0, -1));
    expect(Object.keys(parsedData)).toStrictEqual(['post', 'get']);
    expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');

    await user.click(ContentUtils.path().methodTab('post'));
    expect(ContentUtils.path().operationTab('get')).not.toHaveClass('active');
    expect(ContentUtils.path().operationTab('post')).toHaveClass('active');
    parsedData = _getParsedOasData(relativeJsonPath.slice(0, -1));
    expect(Object.keys(parsedData)).toStrictEqual(['post', 'get']);
    expect(store).toHaveSchemaAt([...relativeJsonPath.slice(0, -1), 'get'], {
      operationId: 'get-user-abc',
      responses: {200: {description: 'OK'}},
      summary: '',
    });
    expect(store).toHaveSchemaAt([...relativeJsonPath.slice(0, -1), 'post'], {
      summary: 'Hello',
    });

    // Changing methods should not change url in path input
    expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');
  });

  it('renders operation component with correct props', async () => {
    expect(_operationPath).toStrictEqual(['paths', '/user/abc', 'post']);
    expect(_operation).toStrictEqual({
      summary: 'Hello',
    });
  });

  describe('path params tests', () => {
    let _node, parsedData;

    it('render other properties', async () => {
      await user.click(ContentUtils.path().pathParamsBtn());
      await user.click(ContentUtils.path().addPathParamRowBtn());
      expect(_parameterPath).toStrictEqual([
        'paths',
        '/user/abc',
        'parameters',
        0,
      ]);
      await user.click(ContentUtils.path().addPathParamRowBtn());
      expect(_parameterPath).toStrictEqual([
        'paths',
        '/user/abc',
        'parameters',
        1,
      ]);
    });

    it('adds path params', async () => {
      // Expect post is selected and add GET too in the current path node
      // so that we can assert later that the methods don't get affected
      // by changing path params
      expect(ContentUtils.path().operationTab('post')).toHaveClass('active');
      await user.click(ContentUtils.path().methodTab('get'));
      await user.click(ContentUtils.path().addOperationBtn('get'));

      expect(
        ContentUtils.path().addOperationBtn('get'),
      ).not.toBeInTheDocument();
      await user.click(ContentUtils.path().methodTab('post'));
      expect(
        ContentUtils.path().addOperationBtn('post'),
      ).not.toBeInTheDocument();

      await user.click(ContentUtils.path().pathParamsBtn());
      await user.click(ContentUtils.path().addPathParamRowBtn());
      expect(ContentUtils.path().pathParamRows()).toHaveLength(1);
      await user.click(ContentUtils.path().addPathParamRowBtn());
      expect(ContentUtils.path().pathParamRows()).toHaveLength(2);
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');

      parsedData = _getParsedOasData(relativeJsonPath.slice(0, -2));
      expect(Object.keys(parsedData)).toStrictEqual(['/user/abc']);

      await user.type(ContentUtils.path().pathParamInputName(0), 'a');
      await user.keyboard('{Tab}');
      // Since changing path params updates current rendered path, rerender
      _node = rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc/{a}');
      parsedData = _getParsedOasData(relativeJsonPath.slice(0, -2));
      expect(Object.keys(parsedData)).toStrictEqual(['/user/abc/{a}']);
      expect(store).toHaveSchemaAt(
        [..._node.relativeJsonPath.slice(0, -1), 'parameters'],
        [
          {in: 'path', name: 'a', required: true, schema: {type: 'string'}},
          {in: 'path', name: '', required: true, schema: {type: 'string'}},
        ],
      );

      expect(ContentUtils.path().operationTab('post')).toHaveClass('active');
      await user.click(ContentUtils.path().methodTab('get'));
      expect(
        ContentUtils.path().addOperationBtn('get'),
      ).not.toBeInTheDocument();
      await user.click(ContentUtils.path().methodTab('post'));
      expect(
        ContentUtils.path().addOperationBtn('post'),
      ).not.toBeInTheDocument();

      await user.type(ContentUtils.path().pathParamInputName(1), 'bc');
      await user.keyboard('{Tab}');
      _node = rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc/{a}/{bc}');
      parsedData = _getParsedOasData(relativeJsonPath.slice(0, -2));
      expect(Object.keys(parsedData)).toStrictEqual(['/user/abc/{a}/{bc}']);
      expect(ContentUtils.path().operationTab('post')).toHaveClass('active');
      await user.click(ContentUtils.path().methodTab('get'));
      expect(
        ContentUtils.path().addOperationBtn('get'),
      ).not.toBeInTheDocument();
      await user.click(ContentUtils.path().methodTab('post'));
      expect(
        ContentUtils.path().addOperationBtn('post'),
      ).not.toBeInTheDocument();
      expect(store).toHaveSchemaAt(
        [..._node.relativeJsonPath.slice(0, -1), 'parameters'],
        [
          {in: 'path', name: 'a', required: true, schema: {type: 'string'}},
          {in: 'path', name: 'bc', required: true, schema: {type: 'string'}},
        ],
      );

      await user.click(ContentUtils.path().removePathParamRowBtn(1));
      _node = rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc/{a}');
      expect(ContentUtils.path().pathParamRows()).toHaveLength(1);
      expect(store).toHaveSchemaAt(
        [..._node.relativeJsonPath.slice(0, -1), 'parameters'],
        [{in: 'path', name: 'a', required: true, schema: {type: 'string'}}],
      );

      await user.click(ContentUtils.path().removePathParamRowBtn(0));
      rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');
      // The tabs exists and both methods do not have "Add operation" buttons
      // meaning they have existing operation body
      expect(ContentUtils.path().operationTab('post')).toHaveClass('active');
      await user.click(ContentUtils.path().methodTab('get'));
      expect(
        ContentUtils.path().addOperationBtn('get'),
      ).not.toBeInTheDocument();
      await user.click(ContentUtils.path().methodTab('post'));
      expect(
        ContentUtils.path().addOperationBtn('post'),
      ).not.toBeInTheDocument();
      expect(store).toHaveSchemaAt(
        [...relativeJsonPath.slice(0, -1), 'parameters'],
        [],
      );
    });

    it('cannot add empty path params', async () => {
      await user.click(ContentUtils.path().pathParamsBtn());

      // Add 3 blank param rows
      await user.click(ContentUtils.path().addPathParamRowBtn());
      await user.click(ContentUtils.path().addPathParamRowBtn());
      await user.click(ContentUtils.path().addPathParamRowBtn());
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');
      // await user.type(ContentUtils.path().pathParamInputName(0), 'a');
      await user.keyboard('{Tab}');
      rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc');

      await user.type(ContentUtils.path().pathParamInputName(1), 'bc');
      await user.keyboard('{Tab}');
      rerender_node();
      expect(ContentUtils.path().pathInput()).toHaveValue('/user/abc/{bc}');

      const parsedData = _getParsedOasData(relativeJsonPath.slice(0, -2));
      expect(Object.keys(parsedData)).toStrictEqual(['/user/abc/{bc}']);
    });
  });

  it('matches snapshot', async () => {
    const {stores, findNode} = initStore({
      paths: {'/user/abc': {post: {summary: 'Hello'}}},
    });
    const node = findNode.path('~1user~1abc');
    stores.uiStore.setActiveNode(node);
    const {asFragment} = render(
      <Path relativeJsonPath={node.relativeJsonPath} node={node} />,
      {
        providerProps: {value: stores},
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
