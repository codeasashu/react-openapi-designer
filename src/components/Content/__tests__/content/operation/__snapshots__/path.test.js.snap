// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path tests matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-1 relative"
  >
    <div
      class="EditorPanel EditorPanel--primary EditorPanel--forms group p-0 flex flex-col relative inset-0"
    >
      <div
        class="flex px-10 mt-10 max-w-6xl justify-between"
      >
        <input
          class="px-2 py-1 font-medium bg-transparent hover:bg-darken-2 focus:bg-darken-2 rounded-lg w-full flex-shrink text-2xl"
          placeholder="Operation Name"
          value="Hello"
        />
      </div>
      <div
        class="px-10 pb-2 max-w-6xl"
      >
        <div
          class="mt-6"
        >
          <div
            class="bp4-control-group flex"
          >
            <button
              class="bp4-button"
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                http://localhost
              </span>
            </button>
            <div
              class="bp4-input-group flex-1"
            >
              <input
                aria-label="path"
                class="bp4-input"
                style="padding-right: 0px;"
                type="text"
                value="/user/abc"
              />
              <span
                class="bp4-input-action"
              />
            </div>
            <button
              class="bp4-button"
              type="button"
            >
              <span
                class="bp4-button-text"
              >
                path params
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col flex-1 w-full m-auto p-10"
      data-tabs="true"
    >
      <ul
        class="mt-6 px-10 flex bp4-simple-tab-list"
        role="tablist"
      >
        <li
          aria-controls="react-tabs-129"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-128"
          role="tab"
        >
          <span
            class="uppercase "
          >
            get
          </span>
        </li>
        <li
          aria-controls="react-tabs-131"
          aria-disabled="false"
          aria-selected="true"
          class="bp4-simple-tab uppercase  selected-tab"
          id="react-tabs-130"
          role="tab"
          tabindex="0"
        >
          <span
            class="uppercase  text-info dark:text-info"
          >
            post
          </span>
        </li>
        <li
          aria-controls="react-tabs-133"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-132"
          role="tab"
        >
          <span
            class="uppercase "
          >
            put
          </span>
        </li>
        <li
          aria-controls="react-tabs-135"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-134"
          role="tab"
        >
          <span
            class="uppercase "
          >
            patch
          </span>
        </li>
        <li
          aria-controls="react-tabs-137"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-136"
          role="tab"
        >
          <span
            class="uppercase "
          >
            delete
          </span>
        </li>
        <li
          aria-controls="react-tabs-139"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-138"
          role="tab"
        >
          <span
            class="uppercase "
          >
            head
          </span>
        </li>
        <li
          aria-controls="react-tabs-141"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-140"
          role="tab"
        >
          <span
            class="uppercase "
          >
            options
          </span>
        </li>
        <li
          aria-controls="react-tabs-143"
          aria-disabled="false"
          aria-selected="false"
          class="bp4-simple-tab uppercase "
          id="react-tabs-142"
          role="tab"
        >
          <span
            class="uppercase "
          >
            trace
          </span>
        </li>
      </ul>
      <div
        aria-labelledby="react-tabs-128"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-get"
        id="react-tabs-129"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-130"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative active block"
        data-testid="operation-tabpanel-post"
        id="react-tabs-131"
        role="tabpanel"
      >
        <div
          class="relative pt-10"
          role="operation-post"
        >
          <div
            data-testid="path-operation-111"
          >
            def
          </div>
        </div>
      </div>
      <div
        aria-labelledby="react-tabs-132"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-put"
        id="react-tabs-133"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-134"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-patch"
        id="react-tabs-135"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-136"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-delete"
        id="react-tabs-137"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-138"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-head"
        id="react-tabs-139"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-140"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-options"
        id="react-tabs-141"
        role="tabpanel"
      />
      <div
        aria-labelledby="react-tabs-142"
        class="FormOasPath__tab-panel rounded-none flex-1 border-l-0 border-r-0 border-b-0 relative"
        data-testid="operation-tabpanel-trace"
        id="react-tabs-143"
        role="tabpanel"
      />
    </div>
  </div>
</DocumentFragment>
`;
